<p style="color:#153462; 
          font-weight: bold; 
          font-size: 30px; 
          font-family: Gill Sans, sans-serif;
          text-align: center;">
          Implementation of NN in Python</p>

<p style="text-align: justify; 
          text-justify: inter-word;
          font-size:17px;">
    The main goal of this notebook is implementing simple artificial neural network from end to end.
</p>

 ### <span style="color:#C738BD; font-weight: bold;">Simple Manual Calculation of Layer</span>

<img src="images\simple-nn-layer.png" alt="simple-nn-layer" style="width: 200px;"/>

<p style="text-align: justify; 
          text-justify: inter-word;
          font-size:17px;">
  Let's assume you have 4 input nodes and 3 output nodes. Now you have $4 * 3 = 12$ weights and 3 bias(to know more about bias look it in nn_terminology notebook)
</p>

In [2]:
input = [1, 2.3, 3, 0.5]

# Random weights
weight_1 = [0.5, 0.4, 0.9, 0.6]
weight_2 = [0.6, 0.2, 0.1, 0.4]
weight_3 = [0.6, 0.1, 0.5, 0.6]

#Bias
bias_1 = 1
bias_2 = 2
bias_3 = 3

# output of each output node
output = [input[0] * weight_1[0] +  input[1] * weight_1[1] + input[2] * weight_1[2]+ input[3] * weight_1[3] + bias_1,
          input[0] * weight_2[0] +  input[1] * weight_2[1] + input[2] * weight_2[2] + input[3] * weight_2[3] + bias_2,
          input[0] * weight_3[0] +  input[1] * weight_3[1] + input[2] * weight_3[2] + input[3] * weight_3[3] + bias_3]
print(output)

[5.42, 3.56, 5.63]


 ### <span style="color:#C738BD; font-weight: bold;">With For loop</span>

In [1]:
inputs = [1, 2.3, 3, 0.5]

# Random weights
weights = [[0.5, 0.4, 0.9, 0.6], [0.6, 0.2, 0.1, 0.4], [0.6, 0.1, 0.5, 0.6]]
biases = [1, 2, 3]

layer_output = []
for weight, bias in zip(weights, biases):
    neuron_output = 0
    for n_input, w in zip(inputs, weight):
        neuron_output += n_input * w
    neuron_output += bias
    layer_output.append(neuron_output)

print(layer_output)

[5.42, 3.56, 5.63]


 ### <span style="color:#C738BD; font-weight: bold;">Implementation with Numpy</span>

In [2]:
import numpy as np

inputs = [1, 2.3, 3, 0.5]

# Random weights
weights = [[0.5, 0.4, 0.9, 0.6], [0.6, 0.2, 0.1, 0.4], [0.6, 0.1, 0.5, 0.6]]
biases = [1, 2, 3]

# Dot product
# The dimention is very important while performing dot product
# inputs has (4, 1) and weights has (3, 4), so meet dot product criteria either do transportation of weight or
# do weights * inputs
layer_output = np.dot(weights, inputs) + biases
print(layer_output)

[5.42 3.56 5.63]


 ### <span style="color:#C738BD; font-weight: bold;">Batches, Layer and Objects</span>

<p style="text-align: justify; 
          text-justify: inter-word;
          font-size:17px;">
  In this sectioin, we are going to feed a batch of inputs. Batch of inputs helps for better
  generalization. When you feed a batch of inputs, the network performs the same operations 
  (using the same weights) on all inputs simultaneously. This parallel processing is what makes
  batching computationally efficient, especially on GPUs. During training, the network updates its
  weights based on the error calculated from the entire batch.The updates aim to improve performance
  across all inputs in the batch, leading to more stable and efficient training.
</p>


In [2]:
# It has size of 4 X 4
batch_of_inputs = [[1, 2, 3, 4],
                   [0.5, 1, 3, 0.4],
                   [4.5, 3, 1, 0.9],
                   [0.5, 0.4, 2, 0.2]]

# Random weights
# It has size of 3 X 4
weights = [[0.5, 0.4, 0.9, 0.6], 
           [0.6, 0.2, 0.1, 0.4], 
           [0.6, 0.1, 0.5, 0.6]]

biases = [1, 2, 3]

In [4]:
import numpy as np

# Performing transpose to make it compatible with matrix multiplication
np.dot(batch_of_inputs, np.transpose(weights))

array([[6.4 , 2.9 , 4.7 ],
       [3.59, 0.96, 2.14],
       [4.89, 3.76, 4.04],
       [2.33, 0.66, 1.46]])

In the output matrix:


\begin{bmatrix}
6.4 & 2.9 & 4.7 \\
3.59 & 0.96 & 2.14 \\
4.89 & 3.76 & 4.04 \\
2.33 & 0.66 & 1.46
\end{bmatrix}


- Each **row** corresponds to an input sample from the `batch_of_inputs`.
- Each **column** corresponds to a neuron in the layer (since there are 3 neurons, we have 3 columns).
- The values in the matrix represent the weighted sum of inputs for each neuron before applying any activation function.

For example:
- The first row `[6.4, 2.9, 4.7]` represents the weighted sum of inputs for each of the 3 neurons for the **first input sample** `[1, 2, 3, 4]`.
- The second row `[3.59, 0.96, 2.14]` represents the weighted sum of inputs for the 3 neurons for the **second input sample** `[0.5, 1, 3, 0.4]`.

Thus, each row in the output represents the pre-activation outputs of all neurons for a single input sample.